### ISSUE TITLE
ZeRO-offload performance issues

### ISSUE THREAD

**Issue by hezhiqian01** [microsoft/DeepSpeed#6695]
**Describe the bug**
The performance drops significantly when using ZeRO-Offload to train Llama2 model.

model | ZeRO Offload | TOTAL TIME | TOTAL/AVG FORWARD TIME | TOTAL/AVG BACKWARD TIME | TOTAL/AVG ADAM TIME
-- | -- | -- | -- | -- | --
llama2-0.3B | true | 150s | 6s/0.02s | 66s/0.26s | 78s/0.31s
llama2-0.3B | false | 23s | 7s/0.02s | 11s/0.04s | 6s/0.02s
llama2-1.2B | true | 490s | 4.8s/0.02s | 174s/0.87s | 310s/1.5s
llama2-1.2B | false | 31s | 11s/0.03s | 10s/0.05s | 16s/0.08s

**To Reproduce**

please run this code
```
import torch
import time
import deepspeed
from transformers import LlamaConfig, LlamaForCausalLM
from torch.utils.data import DataLoader, Dataset

# 伪造数据集
class FakeDataset(Dataset):
    def __init__(self, size=1000, seq_length=128, vocab_size=32000):
        self.size = size
        self.seq_length = seq_length
        self.vocab_size = vocab_size

    def __len__(self):
        return self.size

    def __getitem__(self, idx):
        input_ids = torch.randint(0, self.vocab_size, (self.seq_length,))
        attention_mask = torch.ones(self.seq_length)
        labels = input_ids.clone()
        return {"input_ids": input_ids, "attention_mask": attention_mask, "labels": labels}

# 创建未训练的LLaMA模型配置
# config = LlamaConfig(
#     vocab_size=32000,  # 词汇表大小
#     hidden_size=1024,   # 隐藏层维度
#     num_hidden_layers=16,  # 层数
#     num_attention_heads=32,  # 注意力头的数量
#     intermediate_size=2752,  # 中间层大小
# )

# 1.2B
config = LlamaConfig(
    vocab_size=32000,  # 词汇表大小
    hidden_size=2048,   # 隐藏层维度
    num_hidden_layers=16,  # 层数
    num_attention_heads=32,  # 注意力头的数量
    intermediate_size=8256,  # 中间层大小
)

# 7B
# config = LlamaConfig(
#     vocab_size=32000,  # 词汇表大小
#     hidden_size=4096,   # 隐藏层维度
#     num_hidden_layers=32,  # 层数
#     num_attention_heads=128,  # 注意力头的数量
#     intermediate_size=11008,  # 中间层大小
# )

# 使用自定义配置初始化未经训练的模型
model = LlamaForCausalLM(config)

# 打印模型的参数量
total_params = sum(p.numel() for p in model.parameters())
print(f"Total number of parameters: {total_params}")

# 生成伪造数据集
batch_size = 1
dataset = FakeDataset(size=100, seq_length=512, vocab_size=config.vocab_size)
dataloader = DataLoader(dataset, batch_size=batch_size, shuffle=True)

ds_config = {
    "train_batch_size": batch_size,
    "optimizer": {
        "type": "Adam",
        "params": {
            "lr": 0.001,
            "betas": [0.9, 0.999],
            "eps": 1e-8,
            "weight_decay": 0
        }
    },
    "zero_optimization": {
        "stage": 2,
        "offload_optimizer": {
            "device": "cpu",
        },
        "contiguous_gradients": True,
        "overlap_comm": True
    },
    "fp16": {
        "enabled": False,  
        "loss_scale": 0,  
    }
}

# DeepSpeed 初始化
model_engine, optimizer, _, _ = deepspeed.initialize(
    model=model, 
    model_parameters=model.parameters(), 
    config_params=ds_config
)

def train_one_epoch():
    model_engine.train()
    total_loss = 0
    adam_time = 0
    forward_time = 0
    backward_time = 0
    flag = True
    for batch in dataloader:
        if flag:
            st = time.time()
            inputs = {key: val.to(model_engine.device) for key, val in batch.items()}
            
            outputs = model_engine(**inputs)
            loss = outputs.loss
            forward_time += time.time() - st
            
            st = time.time()
            model_engine.backward(loss)
            backward_time += time.time() - st
            # flag = False
        
        st = time.time()
        model_engine.step()
        adam_time += time.time() - st

        total_loss += loss.item()
    
    # return total_loss / len(dataloader), adam_time / len(dataloader)
    return forward_time, backward_time, adam_time
    
def test_adam_execution_time(epochs=1):
    start_time = time.time()
    
    total_adam_time = 0
    total_forward_time = 0
    total_backward_time = 0
    for epoch in range(epochs):
        # avg_loss, adam_time = train_one_epoch()
        forward_time, backward_time, adam_time = train_one_epoch()
        # print(f"Epoch {epoch + 1}, Average Loss: {avg_loss:.4f}, Average adam time: {adam_time:.4f}")
        print(f"Epoch {epoch + 1}:" + 
              f"\n\tAverage forward time: {forward_time / len(dataloader):.4f}" +
              f"\n\tAverage backward time: {backward_time / len(dataloader):.4f}" + 
              f"\n\tAverage adam time: {adam_time / len(dataloader):.4f}")
        total_adam_time += adam_time
        total_forward_time += forward_time
        total_backward_time += backward_time
    
    end_time = time.time()
    print(f"Total execution time: {end_time - start_time:.2f} seconds," + 
          f"\n\tTotal adam time: {total_adam_time:.4f}," + 
          f"\n\tAverage adam time: {total_adam_time / (epochs * len(dataloader)):.4f}," + 
          f"\n\tTotal forward time: {total_forward_time:.4f}," + 
          f"\n\tAverage forward time: {total_forward_time / (epochs * len(dataloader)):.4f}," + 
          f"\n\tTotal backward time: {total_backward_time:.4f}," +
          f"\n\tAverage backward time: {total_backward_time / (epochs * len(dataloader)):.4f},"
          )

test_adam_execution_time(epochs=2)
```

**Expected behavior**
The ZeRO-offload performance drop too significantly.

**ds_report output**
Please run `ds_report` to give us details about your setup.

**Screenshots**

**System info (please complete the following information):**
 - OS: Ubuntu 22.04 LTS
 - GPU count and types x1 A100
 - Interconnects (if applicable) None
 - Python version 3.9
 - DeepSpeed version 0.15.3
 - pytorch version 2.4.1
 - Any other relevant info about your setup

**Launcher context**
Are you launching your experiment with the `deepspeed` launcher, MPI, or something else?

**Docker context**
Are you using a specific docker image that you can share?

**Additional context**

**Comment by jomayeri** [comment]
The performance with ZeRO-Offload is expected to drop because it is having to do many more transfers between cpu memory and gpu memory. The effect will be greater for the smaller models as you show here because there is less computation to be overlapped to hide the transfer times. Offload is meant to be used with larger models that can't fit into gpu memory.

**Comment by hezhiqian01** [comment]
@jomayeri 
Thanks for your reply.
The PCIe bandwidth is 32GB/s, but in this case, for the llama2-0.3B we have total 0.3B x 2 bytes = 0.6GB and it took 0.22(0.26-.0.04) seconds which the speed is 2.7 GB/s.
Why is there such a large gap between them？

**Comment by jomayeri** [comment]
Looking at your ds_config you are offloading the optimizer states not the parameters. The optimizer multiplier on model size is typically ~12 bytes. That makes the math (0.3*12)/0.22 = 16 GB/s. Perhaps you are only getting 1 DIMMs worth of bandwidth.

**Comment by xylian86** [comment]
@hezhiqian01 Could you please add `torch.cuda.synchronize()` when you measure the time cost for fwd,bwd,step? I rerun the script (and thank you providing it), the fwd time reported from your script is 0.02s while from the profiler is 0.13s. 

And it is recommanded to 1) use fp16 or bf16 (mixed precision training), using fp32 will double the transfer volumes. 2) set larger batch size would lead to better performance for offloading cases.

**Comment by hezhiqian01** [comment]
> Looking at your ds_config you are offloading the optimizer states not the parameters. The optimizer multiplier on model size is typically ~12 bytes. That makes the math (0.3*12)/0.22 = 16 GB/s. Perhaps you are only getting 1 DIMMs worth of bandwidth.

I am a little confused. 
As the paper description that "During the backward propagation on the loss, the gradient for different parameters are computed at different point in the backward schedule. ZeRO-Offload can transfer these gradients for each parameter individually or in small groups to the CPU memory immediately after they are computed. Therefore, only a small amount of memory is required to temporarily hold the gradients on the GPU memory before they are transferred to CPU memory."

The data needed to be transferred is only the gradients? 
Am I right?

**Comment by hezhiqian01** [comment]
> Looking at your ds_config you are offloading the optimizer states not the parameters. The optimizer multiplier on model size is typically ~12 bytes. That makes the math (0.3*12)/0.22 = 16 GB/s. Perhaps you are only getting 1 DIMMs worth of bandwidth

It seems that the "12 bytes" is for the ZeRO-2 not for the zero-offload?

For the ZeRO-Offload the communication Volume is 2M per single direction
![Image](https://github.com/user-attachments/assets/85067049-4ffb-4e2e-9bcb-61c2eba53fe7)
